### PR TITLE
[Fix] Fix #558 enas export

### LINKED ISSUE #558: How to save the model created with an ENAS_Scheduler?
We have this nice tutorial showing how to use the ENAS_Scheduler to train a model.

[ENAS Tutorial](https://autogluon.mxnet.io/tutorials/nas/enas_proxylessnas.html) 

My question is:  How do we save and load trained models created that way?  I've tried 5 different ways and none of them seem to work.  

Based on what I read, the one I think I'm supposed to use is this:

`
mynet.export(modelFile, epoch=numEpochsTrained)
`

However, that produces this error message when I run it:
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/pdb.py", line 1667, in main
    pdb._runscript(mainpyfile)
  File "/usr/lib/python3.6/pdb.py", line 1548, in _runscript
    self.run(statement)
  File "/usr/lib/python3.6/bdb.py", line 434, in run
    exec(cmd, globals, locals)
  File "<string>", line 1, in <module>
  File "/app/src/cli/boundryDetector.py", line 2, in <module>
    import sys,os;
  File "/app/src/code/boundryDetector/../lib/click/src/click/core.py", line 857, in __call__
    return self.main(*args, **kwargs)
  File "/app/src/code/boundryDetector/../lib/click/src/click/core.py", line 810, in main
    rv = self.invoke(ctx)
  File "/app/src/code/boundryDetector/../lib/click/src/click/core.py", line 1099, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/app/src/code/boundryDetector/../lib/click/src/click/core.py", line 613, in invoke
    return callback(*args, **kwargs)
  File "/app/src/cli/boundryDetector.py", line 108, in boundry_detector
    dnn.save(modelFile)
  File "/app/src/code/boundryDetector/./dnn/save.py", line 12, in save
    s.model.export(modelFile, epoch=s.numEpochsTrained)
  File "/usr/local/lib/python3.6/dist-packages/mxnet/gluon/block.py", line 1106, in export
    assert name in aux_names
AssertionError
```

I'm using:
```
autogluon==0.0.11
mxnet==1.6.0
```

Your kind assistance would be greatly appreciated. :)

### PR DESCRIPTION
**Pull request by zhreshold**
*Issue #, if available:*

https://github.com/awslabs/autogluon/issues/558

*Description of changes:*

ENAS network is dynamic so some parameters may not needed during export, remove the check.
Note that this is one-directional serialization for now, since the enas is wrapped dynamic searchable network architecture.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
Can you update the ENAS tutorial and documentation to showcase this export functionality, along with loading the saved model for inference?

**Comment by szha** [comment]
Job PR-559-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-559/1/index.html

File: docs/tutorials/nas/enas_proxylessnas.md [hunk 1]
```diff
@@ -142,6 +142,22 @@ The resulting architecture is:
...
+The trained ENAS network can be saved to disk for future inferences.
+
+```{.python .input}
+mynet.export('enas')
+```
+
+Load it back with mxnet:
+
+```{.python .input}
+mynet.export('enas', 0)
```
Comment (andyDoucette) [inline docs/tutorials/nas/enas_proxylessnas.md:156]: Does it matter what epoch you export?  I imagine you'd need to export the most recently trained epoch, epoch=1 to get the same results as what we just got.  Is that right?

Comment (zhreshold) [reply docs/tutorials/nas/enas_proxylessnas.md:156]: the epoch is for the name appended to the model prefix, not really a constraint anyway

**Comment by Innixma** [comment]
Apologies for not clarifying in my first comment, but could you also add in the tutorial the loaded model making inference predictions? This will help the CI notify us if the save/load functionality is correct. Perhaps an even better test would be to make predictions before saving, then make predictions with the loaded model, and ensure that they are identical.

**Comment by szha** [comment]
Job PR-559-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-559/3/index.html

File: docs/tutorials/nas/enas_proxylessnas.md (hunk 1 above)
Comment (andyDoucette) [reply docs/tutorials/nas/enas_proxylessnas.md:156]: Oh, so it doesn't change the contents of the exported file, just the file name?

Comment (zhreshold) [reply docs/tutorials/nas/enas_proxylessnas.md:156]: Yes, that is correct

File: autogluon/contrib/enas/enas.py [hunk 2]
```diff
@@ -303,6 +303,44 @@ def __repr__(self):
...
+        .. note:: When there are only one input, it will have name `data`. When there
+                  Are more than one inputs, they will be named as `data0`, `data1`, etc.
+
+        Parameters
+        ----------
+        path : str
+            Path to save model. Two files `path-symbol.json` and `path-xxxx.params`
+            will be created, where xxxx is the 4 digits epoch number.
+        epoch : int
+            Epoch number of saved model.
```
Comment (jwmueller) [inline autogluon/contrib/enas/enas.py:319]: Could we add mini code example here showing how to export and reload exported model?

File: autogluon/contrib/enas/enas.py [hunk 3]
```diff
@@ -303,6 +303,44 @@ def __repr__(self):
         reprstr += ')\n'
         return reprstr
 
+    def export(self, path, epoch=0, remove_amp_cast=True):
```
Comment (jwmueller) [inline autogluon/contrib/enas/enas.py:306]: `remove_amp_cast` missing documentation, maybe at least provide link?

File: docs/tutorials/nas/enas_proxylessnas.md (hunk 1 above)
Comment (jwmueller) [reply docs/tutorials/nas/enas_proxylessnas.md:156]: Will this confuse users?  Why not change the argument-name from `epoch` -> `name_prefix`?

Comment (jwmueller) [reply docs/tutorials/nas/enas_proxylessnas.md:156]: Will this confuse users? Maybe should clarify this in the documentation, or even rename `epoch` arg to something like `name_suffix`

File: docs/tutorials/nas/enas_proxylessnas.md [hunk 4]
```diff
@@ -142,6 +142,22 @@ The resulting architecture is:
 mynet.graph
 ```
 
+## Store the trained model as a static network
+
+The trained ENAS network can be saved to disk for future inferences.
+
+```{.python .input}
+mynet.export('enas')
```
Comment (jwmueller) [inline docs/tutorials/nas/enas_proxylessnas.md:150]: What does this tutorial need two `export` calls?

File: docs/tutorials/nas/enas_proxylessnas.md (hunk 1 above)
Comment (andyDoucette) [reply docs/tutorials/nas/enas_proxylessnas.md]: Actually, if it's only a suffix for a name and doesn't change the data saved in any way, then why have the epoch input at all?  Why not just have modelPathBase or modelDirectory as the parameter, and the user can choose what name prefix or directory (with fixed file names under it) to save the model in?  Seems more clear to me to save a model in a directory where I choose the directory name then to have to worry about the fact that it saves the model as 2 files and needs to be loaded as 2 seperate files too.  Just my two cents.

**Comment by zhreshold** [comment]
> Apologies for not clarifying in my first comment, but could you also add in the tutorial the loaded model making inference predictions? This will help the CI notify us if the save/load functionality is correct. Perhaps an even better test would be to make predictions before saving, then make predictions with the loaded model, and ensure that they are identical.

@Innixma Added unittest and I think it's better fit to check for correct functionality in the future.

File: docs/tutorials/nas/enas_proxylessnas.md (hunk 1 above)
Comment (zhreshold) [reply docs/tutorials/nas/enas_proxylessnas.md]: Well, it's somewhat a mxnet's convention that models are saved with four digit epoch number for identification. I agree that in autogluon user doesnot necessarily have to care about that. Changed in the updated version.

**Comment by szha** [comment]
Job PR-559-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-559/6/index.html

**Comment by szha** [comment]
Job PR-559-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-559/7/index.html

**Pull request merged**